### PR TITLE
[Feature] affine eq node: first-class == constraint in affine sets

### PR DESCRIPTION
**Pull request by fabianlim**
Introduces `("eq", expr)` as a first-class AST node for equality constraints in affine sets, replacing the previous parse-time rewrite (from `inter-tile-reduce` / `4bdf137`) that expanded `==` into two `>=` inequalities.

## Changes

- **Tokeniser**: `==` added to `_TOKEN_RE` before `>=` (longest-match)
- **`parse_constraint_list`**: `lhs == rhs` emits `("eq", ("sub", lhs, rhs))` — one node, normalised to `expr == 0` form (mirrors how `>=` constraints normalise to `expr >= 0`)
- **`_eval_node`**: `eq` tag returns the inner expression value
- **`affine_set_contains`**: checks `== 0` for `eq` nodes, `>= 0` for everything else
- **`BoxSet.try_from_affine_set`**: unified pin logic handles `eq` (sets both `lo` and `hi`) alongside existing inequality bounds; duplicated linearise-then-extract code removed

## Spec examples (all parse and evaluate correctly)

- `affine_set<(g) : (g == 0)>`
- `affine_set<(i) : (i == 0)>`
- `affine_set<(p)[c] : (p - c + 2 == 0)>`
- `affine_set<(p)[c, g] : (p + c - 8*g - 3 == 0)>`

## Coordination with PR #69

`test_eq_multi_dim_rejected` in `tests/test_affine.py` is marked with a `TODO` and will need to be revisited when PR #69 lands.

### CODE REVIEW

**Comment by fabianlim** [comment]
CC @WarningRan @nwang-ibm — please review. Note this may need coordination with PR #69 when that branch merges. This PR is needed for regex parsing of affine sets that have `==` in them, which will be the case when we consider inter tile in https://github.com/torch-spyre/ktir-mlir-frontend/pull/23

File: ktir_cpu/parser_ast.py [hunk 1]
```diff
@@ -278,27 +278,27 @@ def parse_expr_list(self) -> List[_Node]:
...
 
-        Each constraint is normalised to ``lhs - rhs >= 0``:
+        Inequality constraints are normalised to ``lhs - rhs >= 0``:
           - ``lhs >= rhs``  → stored as ``("sub", lhs, rhs)``
           - ``lhs <= rhs``  → stored as ``("sub", rhs, lhs)``
 
-        The special case ``expr >= 0`` / ``expr <= 0`` is handled naturally
-        since ``("sub", expr, 0_const)`` evaluates identically to ``expr``.
+        Equality constraints are stored as a first-class node:
+          - ``lhs == rhs``  → stored as ``("eq", lhs, rhs)``
```
Comment (kiszk) [inline ktir_cpu/parser_ast.py:294]:
```suggestion
          - ``lhs == rhs``  → stored as ``("eq", ("sub", lhs, rhs))``
          (normalised to expr == 0 form, similar to inequalities)
```

File: tests/test_affine.py [hunk 2]
```diff
@@ -399,3 +421,66 @@ def test_axis_aligned_becomes_box(self):
...
+        assert s.lo == (0,)
+        assert s.hi == (1,)
+
+    def test_spec_example_symbolic_stays_affineset(self):
+        """Spec example with symbols: affine_set<(p)[c] : (p - c + 2 == 0)>.
+
+        n_syms > 0, so try_from_affine_set rejects it; stays AffineSet.
+        """
+        s = parse_affine_set("affine_set<(p)[c] : (p - c + 2 == 0)>")
+        assert isinstance(s, AffineSet)
```
Comment (kiszk) [inline tests/test_affine.py]: How about adding edge cases?
```suggestion
        assert isinstance(s, AffineSet)

    def test_reject_conflicting_eq_constraints():
        aset = parse_affine_set_raw("affine_set<(d0) : (d0 == 2, d0 == 3)>")
        assert BoxSet.try_from_affine_set(aset) is None

    def test_reject_eq_ineq_conflict():
        aset = parse_affine_set_raw("affine_set<(d0) : (d0 == 2, d0 >= 5)>")
        assert BoxSet.try_from_affine_set(aset) is None

    def test_reject_conflicting_inequalities():
        aset = parse_affine_set_raw("affine_set<(d0) : (d0 >= 5, -d0 + 3 >= 0)>")
        assert BoxSet.try_from_affine_set(aset) is None
```

File: ktir_cpu/affine.py [hunk 3]
```diff
@@ -317,44 +317,58 @@ def intersect(self, other: "BoxSet") -> "BoxSet":
...
+                his[i] = pinned + 1 if his[i] is None else min(his[i], pinned + 1)
+            elif k == 1:
+                # d_i >= pinned
+                los[i] = pinned if los[i] is None else max(los[i], pinned)
+            else:
+                # d_i <= pinned  →  hi is exclusive
+                his[i] = pinned + 1 if his[i] is None else min(his[i], pinned + 1)
+
         if any(v is None for v in los) or any(v is None for v in his):
             return None
```
Comment (kiszk) [inline ktir_cpu/affine.py:500]: Adding detection of contradictions
```suggestion
            return None
        if any(los[i] >= his[i] for i in range(n)):
            return None
```

File: ktir_cpu/affine.py [hunk 4]
```diff
@@ -317,44 +317,58 @@ def intersect(self, other: "BoxSet") -> "BoxSet":
...
-                his[i] = candidate if his[i] is None else min(his[i], candidate)
-            else:
+            i, k = nz[0], coeffs[nz[0]]
+            if abs(k) != 1:
                 return None
+
+            # Solve for d_i: k*d_i + const == 0  →  d_i = -const/k
+            pinned = -const if k == 1 else const
+
+            if c[0] == "eq":
```
Comment (kiszk) [inline ktir_cpu/affine.py]: Adding detection of contradictions
```suggestion
            if c[0] == "eq":
            if los[i] is not None and los[i] != pinned:
                return None
            if his[i] is not None and his[i] != pinned + 1:
                return None
```

**Comment by kiszk** [comment]
Overall, it looks good. Some comments address corner cases.

**Comment by fabianlim** [comment]
@kiszk addressed your comments!

**Review by kiszk — DISMISSED** [review]

**Comment by fabianlim** [comment]
@kiszk i will need to rebase this and incorporate some changes we added for support of dynamic symbols #69

cc: @WarningRan

**Comment by fabianlim** [comment]
rebased it on top of #69. if you could help look again @kiszk @WarningRan

**Review by kiszk — APPROVED** [review]
